Aggiunge capitolo sviluppatore senior (#182)
Il-Libro-Open-Source/book: docs/it/senior.md

# Lo sviluppatore "senior"

L'industria del software si muove alla velocitá della luce.

> Come sviluppatori software siamo costantemente bombardati da notizie di nuove tecnologie, nuovi framework, nuovi linguaggi, nuovi paradigmi, nuovi modelli organizzativi, nuovi modelli di sviluppo, nuovi modelli di deployment, nuovi modelli di testing, nuovi modelli di integrazione, nuovi modelli di delivery, nuovi modelli di monitoraggio, etc.

Author: Non so se é una licenza poetica passabile, ma qui avevo intenzionalmente lasciato la ripetizione per enfatizzare quanto possa essere "estenuante" la continua evoluzione tecnologica...

Suona troppo male?

> Come sviluppatori software siamo costantemente bombardati da notizie di nuove tecnologie, nuovi framework, nuovi linguaggi, nuovi paradigmi, nuovi modelli organizzativi, nuovi modelli di sviluppo, nuovi modelli di deployment, nuovi modelli di testing, nuovi modelli di integrazione, nuovi modelli di delivery, nuovi modelli di monitoraggio, etc.

Reviewer: Forse con le doppie virgolette? Magari aspettiamo feedback da qualcun altro

> Come sviluppatori software siamo costantemente bombardati da notizie di nuove tecnologie, nuovi framework, nuovi linguaggi, nuovi paradigmi, nuovi modelli organizzativi, nuovi modelli di sviluppo, nuovi modelli di deployment, nuovi modelli di testing, nuovi modelli di integrazione, nuovi modelli di delivery, nuovi modelli di monitoraggio, etc.

Reviewer 2: Mi piace l'idea delle virgolette o del corsivo di modo da enfatizzare la voluta ripetizione, secondo me incarnano bene il concetto, li terrei!

> Come sviluppatori software siamo costantemente bombardati da notizie di nuove tecnologie, nuovi framework, nuovi linguaggi, nuovi paradigmi, nuovi modelli organizzativi, nuovi modelli di sviluppo, nuovi modelli di deployment, nuovi modelli di testing, nuovi modelli di integrazione, nuovi modelli di delivery, nuovi modelli di monitoraggio, etc.

Author: il corsivo mi sembra un buon compromesso :) aggiorno!

Grazie <3

Una valanga travolgente di innovazione che non da segni di rallentamento.

É questa l'arma a doppio taglio nella mani di uno sviluppatore software. Da una parte abbiamo a che fare con infinite opportunitá di crescita e il prospetto di non annoiarci mai. Dall'altra parte, siamo sempre alla rincorsa della "next big thing" e, per quanto ci sforziamo di stare al passo con i tempi, siamo sempre inesorabilmente un passo indietro.

In questo contesto cosí dinamico, cosa significa essere uno sviluppatore "senior"?

> Facendo un parallelo, se per esempio dovessimo dare una definizione di un profilo "senior" nell'industria aeronautica, potremmo pensare al ruolo del capitano.

Author: Non sono sicurissimo che valga la pena cambiare. L'idea é proprio quella di far vedere che non si puó fare una analogia con altre industrie e ruoli piú tradizionali che spesso sono piú strutturati del mondo del software...

> Il capitano di un aereo é un profilo che é responsabile di tutti gli aspetti del volo, del velivolo, dei passeggeri e del personale di bordo. É un profilo che può sicuramente contare su diversi anni di studio, una grande esperienza (esplicitamente espressa in ore o miglia di volo). É sicuramente un percorso lungo che richiede tanta dedizione, ma anche un percorso piuttosto chiaro e standardizzato e che probabilmente non ha avuto significative variazioni negli ultimi 40 anni...

Author: Dovrei aver fatto un mega find and replace! grazie <3

## Voglia di crescere

La mentalità di uno sviluppatore senior è cruciale quanto le sue competenze tecniche. Un ingegnere senior deve avere una mentalità incline alla crescita, la volontà di imparare e adattarsi e la propensione al continuo miglioramento.

Siamo tutti consapevoli di quanto velocemente si evolva il mondo del software, lo abbiamo giá detto ed una situazione di disagio che viviamo tutti i giorni. L'unica arma in nostro possesso per poter sopravvivere in questo mondo é quella di essere in grado di imparare velocemente e di adattarci ai cambiamenti.

Imparare peró significa anche essere a proprio agio con i fallimenti e vederli come un’opportunità per imparare. Questo significa talvolta essere disposto a correre dei rischi, provare soluzioni o tecnologie con le quali non abbiamo molta confidenza ed essere pronti a tornare indietro sui propri passi qualora quell'approccio non porti i frutti sperati.

Ogni fallimento peró é un passo in avanti verso il successo, una lezione appresa che ci permetterá di non ripetere gli stessi errori in futuro e di essere piú consapevoli delle scelte che facciamo.

## Essere una leva attiva, saper dire "NO" e debito tecnico

> Uno sviluppatore senior deve avere un ruolo molto proattivo in un’organizzazione. Non dovrebbe isolarsi in una stanza e continuare a sbattere le dita contro la tastiera. Un senior deve sapere quando è il momento di fare domande scomode e prendere in mano la situazione.

Reviewer: ```suggestion
Uno sviluppatore senior dovrebbe avere un ruolo molto proattivo in un’organizzazione. Non dovrebbe isolarsi in una stanza e continuare a sbattere le dita contro la tastiera. Un senior dovrebbe sapere quando è il momento di fare domande scomode e prendere in mano la situazione.
```

Questo a volte significa che un senior dovrebbe anche sapere quando è il momento di dire NO. No a soluzioni rapide e sporche (che alla fine si ritorceranno contro) solo per rispettare una scadenza. No a piani che prevedono solo la costruzione di funzionalità su funzionalità senza considerare l’esperienza dell’utente e la stabilità a lungo termine del prodotto. No a scelte tecniche arbitrarie solo perché “abbiamo sempre fatto così ed è sempre andata bene”.

Dire NO è facile ma al tempo stesso molto difficile. Tutti possono dire NO, questa è la parte facile! Il difficile è argomentare il perché di quel NO, proporre punti di vista alternativi, trovare compromessi e disinnescare bombe a orologeria a breve o lungo termine.

Per riuscire a farlo serve un elevato livello di soft skills: devi saper comunicare in modo efficace usando le parole giuste e scegliere con saggezza le tua battaglie. Non puoi dire sempre NO a tutto o corri il rischio di essere preso per un ostruzionista scorbutico.

> Si tratta piuttosto di saper dire “NO, MA” piuttosto che un secco “NO”: provvedere ad una valida alternativa, seppur non ottima tecnicamente o di requisiti è parte del ruolo. A volte è necessario accettare compromessi e fare scelte che non sono necessariamente le migliori dal proprio punto di vista. Ma è importante essere in grado di riconoscere quando è il momento di dire NO e quando è il momento di accettare un compromesso.

Reviewer: ```suggestion
Si tratta piuttosto di saper dire “NO, MA” piuttosto che un secco “NO”: provvedere ad una valida alternativa, seppur non ottima tecnicamente o di requisiti, è parte del ruolo. A volte è necessario accettare compromessi e fare scelte che non sono necessariamente le migliori dal proprio punto di vista. Ma è importante essere in grado di riconoscere quando è il momento di dire NO e quando è il momento di accettare un compromesso.
```

E dovresti anche essere in grado di farlo di fronte alle persone giuste e al momento giusto… Se fatto bene, questo può avere un forte impatto positivo sull’azienda e sul team. Può portare all’innovazione e creare nuove potenzialità di lavoro o vantaggi competitivi unici.
Questo é un argomento che viene spesso tirato in ballo quando si parla di debito tecnico, ovvero scelte tecniche non ideali che a lungo termine possono portare a problemi di stabilitá e scalabilitá fino al punto di danneggiare il business. Piú il tempo passa e piú il debito tecnicó si accumula e piú sará difficile ripagarlo.

Il debito tecnico é un argomento molto complesso e non esiste una soluzione unica per tutti i casi. A volte é necessario accettare un debito tecnico per poter rispettare una scadenza o per poter testare un'idea. Altre volte é necessario investire tempo per ripagare il debito tecnico accumulato in passato.

> Le capacitá di un senior dovrebbero permettergli di capire quando é il momento di accettare un debito tecnico e quando é il momento di investire tempo per ripagarlo. Inoltre, dovrebbe essere in grado di comunicare in modo efficace i rischi associati al debito tecnico e le conseguenze che questo potrebbe avere sul business. Infine un senior dovrebbe riuscire a negozioare con i product manager e i dirigenti per trovare il giusto compromesso tra debito tecnico e scadenze.

Reviewer: Typo
```suggestion
Le capacitá di un senior dovrebbero permettergli di capire quando é il momento di accettare un debito tecnico e quando é il momento di investire tempo per ripagarlo. Inoltre, dovrebbe essere in grado di comunicare in modo efficace i rischi associati al debito tecnico e le conseguenze che questo potrebbe avere sul business. Infine un senior dovrebbe riuscire a negoziare con i product manager e i dirigenti per trovare il giusto compromesso tra debito tecnico e scadenze.
```

Poiché dovrai parlare con tutti gli interlocutori, dovresti imparare a comunicare concetti tecnici a interlocutori non tecnici in un modo che sia facile da comprendere. Ciò significa anche saper ascoltare i commenti e incorporarli nel tuo lavoro. La capacità di comunicare efficacemente con gli interlocutori è essenziale per garantire che tutti siano sulla stessa lunghezza d’onda e che i progetti siano portati a termine con successo.

Un’altra importante abilità comunicativa per un ingegnere del software è quella di spiegare le incertezze e proporre idee su come affrontarle. Lo sviluppo del software è un processo intrinsecamente incerto e ci saranno sempre incognite e sfide impreviste.

Come ingegnere del software, devi essere in grado di comunicare queste incertezze in modo chiaro e conciso. Devi anche essere in grado di proporre idee su come affrontare queste incertezze e far progredire il progetto.

> Inoltre, gli ingegneri del software devono essere in grado di parlare degli errori e delle lezioni apprese. È importante sottolineare che il fallimento è una parte inevitabile dello sviluppo del software ed è importante essere trasparenti sui fallimenti e sulle lezioni apprese. Ciò richiede capacità di comunicazione efficaci, compresa quella di assumersi la responsabilità degli errori e di comunicare onestamente. Inoltre, è necessario saper comunicare le lezioni apprese dai fallimenti e incorporarle nei progetti futuri. In un certo senso, un errore puó anche essere visto come un investimento a lungo termine. Se questo viene comunicato bene puó essere percepito in modo assolutamente positivo da parte dell'azienda e puó convince a investire ancora di piú nella ricerca di nuove soluzioni e pratiche di sviluppo.

Reviewer: Typo
```suggestion
Inoltre, gli ingegneri del software devono essere in grado di parlare degli errori e delle lezioni apprese. È importante sottolineare che il fallimento è una parte inevitabile dello sviluppo del software ed è importante essere trasparenti sui fallimenti e sulle lezioni apprese. Ciò richiede capacità di comunicazione efficaci, compresa quella di assumersi la responsabilità degli errori e di comunicare onestamente. Inoltre, è necessario saper comunicare le lezioni apprese dai fallimenti e incorporarle nei progetti futuri. In un certo senso, un errore puó anche essere visto come un investimento a lungo termine. Se questo viene comunicato bene puó essere percepito in modo assolutamente positivo da parte dell'azienda e puó convincere ad investire ancora di piú nella ricerca di nuove soluzioni e pratiche di sviluppo.
```

Ti trovi in una posizione unica per capire e sfruttare i punti di forza del team e per coinvolgere le persone giuste al momento giusto. Lavorando a stretto contatto con la direzione, puoi assicurarti che il tuo lavoro sia in linea con gli obiettivi aziendali generali e che gli sforzi del tuo team siano incanalati nella giusta direzione.

La pianificazione è un’altra abilità determinante per un ingegnere del software. Richiede di suddividere i progetti di grandi dimensioni in attività più piccole e gestibili, di stimare il tempo necessario per ciascuna attività e di stabilirne la priorità. Una pianificazione efficace contribuisce a garantire che i progetti siano portati a termine in tempo e nel rispetto del budget.

L’organizzazione degli incontri, come le riunioni in piedi, le revisioni informali e le retrospettive, possono offrire l’opportunità di condividere i progressi, identificare i problemi e pianificare il futuro. Come ingegnere del software, devi contribuire a guidare questi incontri e assicurarti che siano produttivi ed efficienti.

> Ovviamente gli sviluppatori senior non devono e non possono sostituire il ruolo dei project manager o dei product manager, ma piuttosto devono supportare questi ruoli e fornire la loro prospettiva tecnica riuscire a mantenere gli obiettivi, i piani e l’esecuzione in linea con le aspettative del team tecnico.

Reviewer: ```suggestion
Ovviamente gli sviluppatori senior non devono e non possono sostituire il ruolo dei project manager o dei product manager, ma piuttosto devono supportare questi ruoli e fornire la loro prospettiva tecnica per riuscire a mantenere gli obiettivi, i piani e l’esecuzione in linea con le aspettative del team tecnico.
```

## Autonomia e attenzione alla consegna

> Le ultime due soft skills su cui si soffermeremo sono l’autonomia e l’attenzione alla consegna.

Reviewer: Typo
```suggestion
Le ultime due soft skills su cui ci soffermeremo sono l’autonomia e l’attenzione alla consegna.
```

A volte gli sviluppatori senior si trovano in una posizione unica per contribuire in modo significativo a progetti che richiedono una ricerca approfondita e tanto impegno. Si tratta di progetti fortemente tecnici come quelli relativi all’ottimizzazione delle prestazioni, alla significativa ristrutturazione del codice ed alla riprogettazione dell’architettura. Spesso si tratta di progetti molto importanti per l’azienda, perché possono sbloccare l’innovazione e generare nuove potenzialità di lavoro.

> In base alla mia esperienza, non è raro che in questi progetti un singolo collaboratore sia in grado di dedicare una certa quantità di tempo a concentrarsi e a compiere progressi significativi in un lasso di tempo molto breve, portando avanti una prima grossa fetta della ricerca. Ma anche in queste situazioni è importante evitare a tutti i costi il rischio di isolarsi dal resto del team.

Reviewer: Che ne diresti di rielaborare questo pensiero? Lo trovo un tantino intricato

Se stai lavorando a un progetto di questo tipo, assicurati di comunicare regolarmente con la direzione e altri colleghi senior per tenere sotto controllo i tuoi progressi, ottenere un riscontro utile e condividere le tue conoscenze.

Ma la capacità di fornire valore lavorando in autonomia non è utile solo nei progetti di ricerca, ma è anche una buona abilità da padroneggiare in generale. Nell’industria del software, è più comune lavorare con cose che non conosciamo piuttosto che con cose che conosciamo!

Ci sono sempre nuovi ambiti da esplorare, nuove sfide da affrontare e nuovi strumenti con cui confrontarsi. Pertanto, è importante essere in grado di imparare rapidamente e di aggiornarsi velocemente quando necessario. Non dovresti aspettare che siano gli altri a insegnarti e a dirti come fare certe cose, ma dovresti essere in grado di fare alcuni progressi per conto tuo, ad esempio consultando e comprendendo la documentazione e il codice esistente.

> Se hai già una vasta competenza, dovrebbe essere quasi istintivo aggiungerne altre quando necessario. Naturalmente, è giusto chiedere aiuto se ti senti bloccato, ma è anche importante cercare di fare tutto il possibile per progredire da solo, in modo da costruire il maggior contesto possibile e massimizzare il risultato del tempo speso a ricevere aiuto da altri.

Reviewer: Typo 
```suggestion
Se hai già una vasta competenza, dovrebbe essere quasi istintivo aggiungerne altra quando necessario. Naturalmente, è giusto chiedere aiuto se ti senti bloccato, ma è anche importante cercare di fare tutto il possibile per progredire da solo, in modo da costruire il maggior contesto possibile e massimizzare il risultato del tempo speso a ricevere aiuto da altri.
```

> In teoria, un ingegnere del software senior non ha bisogno di troppe indicazioni. È in grado di capire da solo ciò che è necessario per portare avanti le cose e di coinvolgere gli altri quando è necessario.

Reviewer: Per evitare la ripetizione
```suggestion
In teoria, un ingegnere del software senior non ha bisogno di troppe indicazioni. È in grado di capire da solo ciò che è necessario per portare avanti le cose e di coinvolgere gli altri nel momento opportuno.
```

Un altro aspetto riguarda la negoziazione delle aspettative e la comprensione di ciò che significa avere successo nell’ambiente attuale.

È bene essere chiari con la direzione e con il resto del team tecnico e definire le caratteristiche del successo individuale e di squadra. Se non sai cosa ci si aspetta da te e dal resto del team, anche se pensi di fare la cosa giusta, potresti finire per non fornire il valore che ci si aspetta da te; allo stesso modo, potresti non essere in grado di aiutare il tuo team ad avere successo.

> Avere un simile atteggiamento di schiettezza e chiarezza di solito mantiene l’umore più positivo e aiuta a evitare le delusioni. Certo, non è sempre possibile evitare tutte le delusioni o i fallimenti. Perciò, quando il problema si presenta, potresti essere tentato di dare la colpa a qualcun altro o al sistema. Ma è invece più maturo chiedersi cosa si sarebbe potuto fare meglio, o ancora meglio: cosa si potrebbe fare in futuro per evitare che problemi simili si ripresentino.

Reviewer: Contrasto di superlativi
```suggestion
Avere un simile atteggiamento di schiettezza e chiarezza di solito mantiene l’umore più alto e aiuta a evitare le delusioni. Certo, non è sempre possibile evitare tutte le delusioni o i fallimenti. Perciò, quando il problema si presenta, potresti essere tentato di dare la colpa a qualcun altro o al sistema. Ma è invece più maturo chiedersi cosa si sarebbe potuto fare meglio, o ancora meglio: cosa si potrebbe fare in futuro per evitare che problemi simili si ripresentino.
```

## Come posso crescere?

Se sei un ingegnere del software e vuoi aumentare le tue competenze e diventare un membro senior del tuo team, potresti chiederti quali sono i passi da intraprendere per raggiungere questo obiettivo.

Beh, non temere! Ecco 4 idee (più alcune extra) che credo ti aiuteranno nel tuo percorso.

### Scendere giú di un livello

Innanzitutto, non limitarti a sfiorare la superficie delle tecnologie con cui stai lavorando. Scava più a fondo e impara a conoscere gli strati sottostanti. Probabilmente hai costruito un sito web o un'API, ma ti sei mai chiesto come funziona il protocollo HTTP? O anche come funziona il protocollo TCP e cosa succede per stabilire una connessione? Apprezzare questi dettagli ti darà una comprensione molto più ricca delle tecnologie con cui lavori ogni giorno.

Sì, se cominci a scavare diventa un pozzo senza fondo... Quindi come fare ad addentrarsi più a fondo senza perdersi?

> La mia regola generale (che ho rubato/appreso dal mio amico [Roberto](https://www.linkedin.com/in/gambuzzi/)) è quella di scendere solo un livello più in basso, partendo dalle tecnologie con cui si ha già familiarità.

Reviewer: Più leggibile?
```suggestion
La mia regola generale (che ho rubato/appreso dal mio amico [Roberto](https://www.linkedin.com/in/gambuzzi/)) è quella di scendere solo di un livello più in basso alla volta, partendo dalle tecnologie con cui si ha già familiarità.
```

Ti serve qualche esempio?

Hai già trascorso un po’ di tempo con OAuth e OpenID Connect? Probabilmente vorrai sapere cosa c’è dentro un JWT. Conosci il flusso del codice di autorizzazione OAuth e cosa succede con tutti i reindirizzamenti? Sai come funzionano gli algoritmi di firma come RSA e HMAC?

Tutti questi sono argomenti che puoi esplorare per consolidare la tua conoscenza di questo particolare ambito. Ti permetteranno di esplorare concetti più generali che potresti riutilizzare in altri ambiti. Tutti costituiscono dei validi passi in avanti per ampliare le tue conoscenze e diventare un ingegnere più completo.

Ed il modo migliore per esplorare questi piani é quello di costruire qualcosa. Puoi capire veramente qualcosa se riesci a costruirne un piccolo prototipo. Ad esempio potresti provare a decodificare un token JWT senza usare una libreria. Saresti poi in grado di andare oltre e implementare anche l’algoritmo di verifica della firma? Magari poresti anche provare a implementare un server OAuth2.0 per capire meglio come funziona il protocollo.

> Costruire prototipi è un ottimo modo per memorizzare alcuni concetti e mettere davvero alla prova la vostra conoscenza. Naturalmente, costruire richiede uno sforzo maggiore, quindi scegli con saggezza gli esercizi di prototipazione: non è possibile reimplementare oltre 60 anni di ingegneria del software solo per il gusto di imparare!

Reviewer: ```suggestion
Costruire prototipi è un ottimo modo per memorizzare alcuni concetti e mettere davvero alla prova la vostra conoscenza. Naturalmente, costruire richiede uno sforzo maggiore, quindi scegli con saggezza gli esercizi di prototipazione: non è possibile reimplementare oltre sessant'anni di ingegneria del software solo per il gusto di imparare!
```

Allo stesso modo, non scartare il valore dell’insegnamento agli altri. Anche i membri più anziani del team possono beneficiare delle tue conoscenze e della tua esperienza. Se fai coppia con altri, puoi condividere le tue competenze e aiutare tutti a crescere e a svilupparsi.

Ho imparato tantissimi trucchi lavorando in coppia con altre persone. Anche piccole cose, come la configurazione dell’editor o del terminale. Lo so: non tutti amano la programmazione in coppia. Se questo è il caso, non preoccuparti! Ci sono altri approcci da seguire che possono comunque fornirti un feedback prezioso e aiutarti a imparare dai tuoi colleghi. Le revisioni interattive del codice e le sessioni di dimostrazione sono ottime alternative che consentono di condividere il proprio lavoro e ricevere riscontri senza dover lavorare insieme in tempo reale.

> Alla fine, la cosa più importante è rimanere aperti a nuove idee e approcci. Continuando a imparare e a crescere, sarei sulla buona strada per diventare uno sviluppatore di software senior. Quindi, perché non provare a fare un po' di pair programming e vedere cosa riesci a imparare?

Reviewer: Typo?
```suggestion
Alla fine, la cosa più importante è rimanere aperti a nuove idee e approcci. Continuando a imparare e a crescere, sarestii sulla buona strada per diventare uno sviluppatore di software senior. Quindi, perché non provare a fare un po' di pair programming e vedere cosa riesci a imparare?
```

> # Lo sviluppatore "senior"

Reviewer 2: ```suggestion
---
layout: default
title: Sviluppatore "senior"
nav_order: 8
---

# Sviluppatore "senior"
```

L'industria del software si muove alla velocitá della luce.

Come sviluppatori software siamo costantemente bombardati da notizie di _nuove_ tecnologie, _nuovi_ framework, _nuovi_ linguaggi, _nuovi_ paradigmi, _nuovi_ modelli organizzativi, _nuovi_ modelli di sviluppo, _nuovi_ modelli di deployment, _nuovi_ modelli di testing, _nuovi_ modelli di integrazione, _nuovi_ modelli di delivery, _nuovi_ modelli di monitoraggio, etc.

Una valanga travolgente di innovazione che non da segni di rallentamento.

È questa l'arma a doppio taglio nella mani di uno sviluppatore software. Da una parte abbiamo a che fare con infinite opportunitá di crescita e il prospetto di non annoiarci mai. Dall'altra parte, siamo sempre alla rincorsa della "next big thing" e, per quanto ci sforziamo di stare al passo con i tempi, siamo sempre inesorabilmente un passo indietro.

> In questo contesto cosí dinamico, cosa significa essere uno sviluppatore "senior"?

Reviewer 3: Correggere accenti acuti: es così, velocità ecc (vale per tutto il capitolo)

Inoltre (un puntiglio) lo metto qui generico per tutto il file:
etc. o ecc. è un'abbreviazione e vuole il punto (tendenzialmente in italiano metto ecc. ma credo che vada bene anche etc., la cosa importante è rimanere costanti)

Author: vediamo se riesco a trovare un modo semplice per risolvere tutti questi (benedetta tastiera UK su mac...)

Author: Fatto grazie a M$ wordz! 🎉🎉

**Word is my new markdown editor!** 💪

Facendo un parallelo, se per esempio dovessimo dare una definizione di un profilo "senior" nell'industria aeronautica, potremmo pensare al ruolo del capitano.

Il capitano di un aereo è un profilo che è responsabile di tutti gli aspetti del volo, del velivolo, dei passeggeri e del personale di bordo. È un profilo che può sicuramente contare su diversi anni di studio, una grande esperienza (esplicitamente espressa in ore o miglia di volo). È sicuramente un percorso lungo che richiede tanta dedizione, ma anche un percorso piuttosto chiaro e standardizzato e che probabilmente non ha avuto significative variazioni negli ultimi 40 anni...

Possiamo trovare una definizione simile anche in ambito software?

_Dipende!_

In questo capitolo cercheremo di esplorare cosa significa essere senior in ambito software. Discuteremo cosa puó e deve offrire un senior al suo team e alla sua azienda, dagli aspetti più tecnici (hard skills) a quelli più umani (soft skills).

> Ma prima di imbarcarci in questo viaggio, è importante puntualizzare che non esiste una definizione canonica di "senior": ogni azienda, ogni team e ogni individuo ha la propria definizione.

Reviewer 3: ogni azienda, ogni team e ogni individuo ha la propria definizione.
io direi ogni azienda, ogni team e ogni individuo `hanno` la propria definizione. -> indica la pluralità delle entità espansa alla pluralità delle definizioni

## Senior? Si, ma quanto?

Immaginiamo per un attimo di essere il CTO di un'azienda in forte crescita. Abbiamo costruito un prodotto grazie al supporto di un'azienda di consulenza esterna, ma adesso è fondamentale portare lo sviluppo _in-house_ e costruire un team interno per poter rendere l'azienda stabile e raggiungere i prossimi obiettivi di crescita.

Quello che ci serve in questa fase é un team di sviluppatori senior che possa costituire la base attorno alla quale costruire e far crescere un team più ampio e variegato di professionisti del software.

Iniziamo a pensare a come scrivere una _job spec_ - ovvero una descrizione dell'offerta e del ruolo - per la ricerca di questi sviluppatori senior. Cosa dovremmo scrivere? Cosa dovremmo cercare? Come definiamo il profilo che stiamo cercando?

> Non abbiamo ancora scritto neanche il titolo del profilo ricercato e siamo giá in preda ad un blocco creativo. Forse è il caso di andare a sbirciare cosa fanno le aziende concorrenti o altre aziende leader nel mercato del software...

Reviewer 3: Non abbiamo ancora scritto neanche il titolo 
toglierei la doppia negazione -> Non abbiamo ancora scritto il titolo ... e siamo già

> Dopo una breve ricerca online, ci troviamo ad avere compilata una lista di possibili ruoli che sembrano essere sinonimi di _senior_: Sviluppatore Senior, Sviluppatore _Staff_, Sviluppatore _Principal_, Architetto del software, Sviluppatore Mago, Sviluppatore Barbabianca, Sviluppatore Unicorno, Rockstar, Ninja, Guru...

Reviewer 3: Qui metterei in risalto Barbabianca allo stesso modo si Principal ecc

Author: Intendi in corsivo?

Nella mia carriera ho avuto la fortuna di lavorare con dei colleghi giovani ma brillanti, in grado di comprendere a fondo problemi tecnici e di business e in grado di dare un grandissimo contributo al team in cui lavoravano.

Mi è anche capitato di lavorare con colleghi con tanti anni di esperienza e anche grosse aziende sul CV che invece si sono limitati a fare il minimo indispensabile e, quando coinvolti nella risoluzione di problemi complessi, mostravano serie lacune sia tecniche che di comprensione del business.

Certo, queste sono estremizzazioni e non sono sicuramente la norma. Il punto è che non bisogna assolutamente puntare né agli anni di esperienza né all'età, ma piuttosto mirare alle competenze acquisite e alla capacità (e volontà) di continuare ad acquisirne di nuove.

## Le competenze

Se non è una questione di anni di esperienza sul campo su cosa dobbiamo concentrarci?

Se definiamo un senior come qualcuno in grado far progredire progetti e persone, ci saranno sicuramente delle competenze necessarie al fine di poter raggiungere questi obiettivi.

> Queste competenze non sono esclusivamente tecniche ma anche non-tecniche o di comunicazione. Uno sviluppatore senior deve saper familiarizzare con piú linguaggi di programmazione, framework, database, architetture software, etc, ma deve anche avere ottime capacitá comunicative, di leadership, di business e problem-solving.

Reviewer 3: Qui `non-tecniche` non è propriamente italiano credo. non-tecniche -> `non-tecniche` oppure "non tecniche"

A volte le _skill_ non tecniche possono risultare anche più importanti di quelle prettamente tecniche, perché non importa solamente quello che sai, ma soprattutto come tu sia in grado di trasmetterlo permettendo agli altri di sfruttare il tuo sapere, come condividere la tua esperienza, sbloccare situazioni difficili e rendere il gruppo di lavoro autosufficiente e produttivo.

L’attitudine al comando è altrettanto importante, in quanto è responsabilità del senior quella di fare da tutor ai junior della squadra, dirigere i vari gruppi del progetto e prendere le scelte tecniche più difficili. Ed è suo anche il compito di motivare la squadra, assegnare i compiti in modo efficace e fare da guida e supporto all’occorrenza.

Un’altra capacità indispensabile per lo sviluppatore senior è la propensione alla risoluzione dei problemi. Uno sviluppatore senior deve essere in grado di pensare in modo critico, analizzare problemi complessi e trovare soluzioni creative. Dovrebbe essere a suo agio all’interno di situazioni poco chiare e non ancora definite, pronto ad adattarsi ad ogni nuovo sviluppo.

In fin dei conti, l’ingegneria del software non è altro che risoluzione di problemi!

> Tuttavia i problemi non sono sempre di natura tecnica, specie all’interno delle organizzazioni più strutturate, dove sono presenti interlocutori con esperienze diverse e di differente formazione, e i problemi possono essere comunicati nel modo sbagliato e mutare di frequente. Per questo diventa fondamentale contribuire alla comprensione dei problemi che si stanno riscontrando, formulare e condividere gli obiettivi e guidare la squadra di lavoro nella giusta direzione; competenze cruciali, queste, che possono avere un impatto enorme.

Reviewer 3: > Tuttavia i problemi non sono sempre di natura tecnica, specie all’interno delle organizzazioni più strutturate, dove sono presenti interlocutori con esperienze diverse e di differente formazione, e i problemi possono essere comunicati nel modo sbagliato e mutare di frequente.

Questa frase mi perde tanto in questo punto: ...esperienze diverse e di differente formazione, e i problemi possono essere comunicati...
forse un `;`?

Bisogna trovare il modo di lasciare margine per commettere errori, imparare dai fallimenti e provare approcci alternativi. La strada per il successo raramente è una linea retta…

## Andare a caccia di bug

È risaputo che il software privo di bug non esiste, quindi non proviamo neanche a pensare che uno sviluppatore senior possa scrivere codice senza bug!

Benché i bug sono sempre dietro l'angolo, uno sviluppatore senior deve sapere cosa fare per ridurre al minimo il rischio di creare dei bug. Analogamente lo sviluppatore senior deve sapere come individuare i bug il più presto possibile, possibilmente prima che lo facciano gli utenti.

Questo comporta sviluppare tutta una serie di competenze quali: comprendere ed approfondire i resoconti degli utenti, scrivere diversi tipi di test (unitari, di integrazione, end-to-end), individuare ed analizzare i casi limite, strutturare i log in modo che siano più facili da leggere e ricercare in caso di errori, riuscire a leggere ed interpretare varie metriche e dashboard.

> Un senior dovrebbe essere in grado di fare una grande differenza e di aiutare l’intero team a scrivere un software migliore, illustrando ed educando alle migliori pratiche per il test del software.

Reviewer 3: La butto lì: linkone al capitolo sul testing? @Cadienvan

Reviewer 2: @nicolaerario ci potrebbe stare!

Author: C'é un modo ben preciso per creare link tra capitoli?

Reviewer 3: Metodo classico dei link con path relativo (vscode dovrebbe dare anche una mano).
Direi: `[testo del link](testing.md)` , dovrebbe andar bene;
Dovessimo linkare una risorsa in diversa lingua (ad esempio un link ad una pagina che, per assurdo, abbiamo solo in inglese), sarebbe: `[testo del link](../en/testing.md)`

## Promuovere le buone pratiche

Uno sviluppatore senior dovrebbe aver sviluppato un certo grado di affinitá con diversi linguaggi di programmazione e dovrebbe conoscere diversi paradigmi e le loro differenze. Ad esempio la programmazione ad oggetti (OOP) e quella funzionale; uno stile dichiarativo rispetto ad uno imperativo; differenze tra linguaggi compilati e interpretati. Possedere questo tipo di competenze puó essere utile per risolvere lo stesso problema utilizzando strumenti e tecniche differenti e potendo cosí scegliere la soluzione che offre i compromessi piú idonei per il problema in questione e per il team con cui si lavora.

Similarmente è importante avere una buona conoscenza dei design pattern al fine di riuscire a strutturare il codice in modo piú flessibile, rendendo piú semplici potenziali evoluzioni future.

Il codice è un elemento che dovrà evolversi con il cambiare delle esigenze del business ed è quindi importante progettare tenendo conto dell’evoluzione. I design pattern non sono una panacea ed a volte è facile abusarne e creare troppi livelli di astrazione. Un buon senior dovrebbe riuscire a trovare il giusto equilibrio tra semplicitá e flessibilitá.

> Bisogna conoscere le buone pratice come i principi del "Clean code" ma è anche importante non essere dogmatici, ovvero applicare tutti i design pattern e le buone pratiche solo perché possiamo farlo. Bisogna piuttosto chiedersi: "qual è il vantaggio di applicare uno specifico pattern ad un dato problema?". Oppure "qual è il costo in termini di complessitá aggiuntiva?".

Reviewer 3: pratice -> pratic`h`e

## Quali tipi di soft skills?

Abbiamo discusso le competenze tecniche che ti daranno un vantaggio nella tua carriera e ti permetteranno di essere riconosciuto come sviluppatore senior. Ma queste da sole non ti porteranno molto lontano. È arrivato il momento di discutere delle fantomatiche "soft skills".

Se hai ottenuto quel colloquio per quella posizione a cui tieni tanto grazie alle tue competenze tecniche, riuscirai ad eccellere e a farti assumere se sarai in grado di dimostrare di avere delle solide soft skills.

Ok, ok, ma quali sono queste soft skills? Come faccio a svilupparle? Come faccio a dimostrare di averle?

È possibile trovare online liste infinite di soft skills che includono intraprendenza, capacitá di comunicazione, cura per i dettagli, attitudine al "problem solving", leadership, negoziazione, gestione del tempo, e chi piú ne ha piú ne metta.

Di seguito discuteremo alcune soft skills che ritengo essere fondamentali per un senior. Non è una lista da considerare esaustiva ma piuttosto un punto di partenza per capire in quali ambiti è possibile misurarsi e migliorarsi.

> 

Reviewer 3: Da qui in poi (fino a 288) si tratta di una lista di soft? Metterei H3 (###)

Per riuscire a farlo serve un elevato livello di soft skills: devi saper comunicare in modo efficace usando le parole giuste e scegliere con saggezza le tua battaglie. Non puoi dire sempre NO a tutto o corri il rischio di essere preso per un ostruzionista scorbutico.

Si tratta piuttosto di saper dire “NO, MA” piuttosto che un secco “NO”: provvedere ad una valida alternativa, seppur non ottima tecnicamente o di requisiti, è parte del ruolo. A volte è necessario accettare compromessi e fare scelte che non sono necessariamente le migliori dal proprio punto di vista. Ma è importante essere in grado di riconoscere quando è il momento di dire NO e quando è il momento di accettare un compromesso.
E dovresti anche essere in grado di farlo di fronte alle persone giuste e al momento giusto… Se fatto bene, questo può avere un forte impatto positivo sull’azienda e sul team. Può portare all’innovazione e creare nuove potenzialità di lavoro o vantaggi competitivi unici.

> Questo è un argomento che viene spesso tirato in ballo quando si parla di debito tecnico, ovvero scelte tecniche non ideali che a lungo termine possono portare a problemi di stabilitá e scalabilitá fino al punto di danneggiare il business. Piú il tempo passa e piú il debito tecnicó si accumula e piú sará difficile ripagarlo.

Reviewer 3: debito tecnicó -> debito tecnic`o`

Potresti ad esempio creare dei contenuti per dare le tue prime impressioni dopo aver provato un nuovo linguaggio di programmazione o un framework. Cosa ti è piaciuto? C’è stato qualcosa che ti ha confuso? Cosa ti è mancato rispetto alla tua precedente esperienza con altri strumenti simili?

C'é un trucco: devi farlo diventare un’abitudine. Impegnati a seguire un programma regolare e svilupperai disciplina e coerenza nella creazione di contenuti. Inoltre, più creerai, più affinerai le tue capacità di comunicazione. E lo ripeto ancora una volta: la comunicazione è un’abilità fondamentale per qualsiasi sviluppatore senior.

A proposito di comunicazione, è qui che la creazione di contenuti brilla. Come ingegneri, amiamo perderci nei dettagli tecnici, ma spiegare concetti complessi a interlocutori non tecnici può essere difficile. Creando contenuti, imparerai a comunicare i tuoi concetti in modo chiaro e coinvolgente per un pubblico più ampio. Inoltre, avrai l’opportunità di ricevere un parere sui tuoi contenuti e, se presterai attenzione, potrai migliorare ulteriormente le tue capacità comunicative.

> Hai mai sentito parlare di [atomic essays](https://www.ship30for30.com/post/how-to-write-an-atomic-essay-a-beginners-guide) (saggi atomici)? Si tratta di suddividere le tue idee in parti di contenuto più brevi e mirate. Questo approccio può rendere i tuoi contenuti più coinvolgenti e più facili da assimilare, soprattutto su piattaforme social come X, dove la brevità è fondamentale. Se non sai che tipo di contenuti creare, questo potrebbe essere un buon formato per iniziare.

Reviewer 3: benedetto Elon. Conviene rendere  la X in `social come X` un link? giusto per dargli uno spessore maggiore di 1 char?

> Se non sai su cosa creare contenuti, ecco il mio consiglio: ogni giorno, alla fine della tua giornata lavorativa, scrivi 1 cosa nuova che pensi di aver imparato durante la giornata. Alla fine della settimana lavorativa, rivedi i tuoi punti. Sono sicuro che per almeno uno di questi potrebbe essere un argomento da condividere e sul quale potresti creare dei contenuti.

Reviewer 3: scrivi 1 cosa nuova -> scrivi `una` cosa nuova
per coerenza con ..."uno di questi"... dopo

La creazione di contenuti può aiutarti a diventare uno sviluppatore più esperto, affermandoti come personaggio di rilievo in un dato ambito, sviluppando le tue capacità di comunicazione e contribuendo alla comunità. Quindi, vai e condividi le tue conoscenze con il mondo.

Ti sorprenderai dell’impatto che avrà su di te e sugli altri!

### Altre idee

Ecco alcuni altri consigli in ordine sparso che puoi aggiungere a quelli discussi sopra.

Innanzitutto, cerca sempre di mantenere un atteggiamento positivo. Quando le cose si fanno difficili, è facile scoraggiarsi e perdere di vista il quadro generale. Ma ricorda che con tempo e risorse sufficienti, i team possono superare qualsiasi sfida. Quando manca questo lusso, sono certo che esistono compromessi accettabili che possono soddisfare le esigenze di tutti.

> Un’altra idea interessante per aggiornarsi ed evitare di restare troppo indietro in fatto di tecnologia o di stili di programmazione e quella di allenarsi con le "coding challenges", ovvero esercizi di programmazione come quelli che puoi trovare su [HackerRank](https://www.hackerrank.com/), [CodeWars](https://www.codewars.com/) o [Exercism](https://exercism.org/). Questi problemi spesso richiedono di scoprire o rispolverare specifici algoritmi o strutture dati. Puoi inoltre usare questi esercizi per esplorare nuovi linguaggi o metodologie. Sforzati di sperimentare diversi modi per risolvere i problemi. È un ottimo esercizio per le tue capacità di risoluzione dei problemi.

Reviewer 3: > Un’altra idea interessante per aggiornarsi ed evitare di restare troppo indietro in fatto di tecnologia o di stili di programmazione e quella

-> Un’altra idea interessante per aggiornarsi ed evitare di restare troppo indietro in fatto di tecnologia o di stili di programmazione `è` quella

## Come faccio a vendermi come sviluppatore senior?

Se penso di essere uno sviluppatore senior, cosa posso fare per essere riconosciuto come tale?

Innanzitutto, inizia ad assumerti maggiori responsabilità. Cerca opportunità per dirigere progetti o fare da tutor a ingegneri junior. Condividi le tue conoscenze e la tua esperienza con gli altri e fornisci loro una guida.

In secondo luogo, cerca di avere un impatto al di là del tuo team o progetto attuale. Partecipa a iniziative interfunzionali, contribuisci a progetti open-source o intervieni a eventi di settore. Questo tipo di attività può aiutarti ad affermarti come guida intellettuale nel tuo settore e a dimostrare la tua esperienza a un pubblico più ampio.

In terzo luogo, concentrati sullo sviluppo delle tue soft skills. In qualità di ingegnere senior, dovrai comunicare efficacemente con interlocutori tecnici e non, condurre riunioni e negoziare efficacemente. Pertanto, lavora per migliorare le capacità di comunicazione, di leadership e di risoluzione dei conflitti per diventare un membro completo e rispettato del tuo team.

> Un altro modo per dimostrare la tua competenza è condividere le tue conoscenze con gli altri. Scrivi post su blog tecnici, registra video didattici o intervieni a eventi interni o esterni per mostrare la tua esperienza e dimostrare la tua volontà di aiutare gli altri a imparare.

Reviewer 3: > didattici o intervieni a eventi interni o esterni

didattici o intervieni a eventi interni `ed` esterni per alleviare la `o`sità

# Lo sviluppatore "senior"

L'industria del software si muove alla velocitá della luce.

Come sviluppatori software siamo costantemente bombardati da notizie di nuove tecnologie, nuovi framework, nuovi linguaggi, nuovi paradigmi, nuovi modelli organizzativi, nuovi modelli di sviluppo, nuovi modelli di deployment, nuovi modelli di testing, nuovi modelli di integrazione, nuovi modelli di delivery, nuovi modelli di monitoraggio, etc.

> Una valanga travolgente di innovazione che non da segni di rallentamento.

Author: Personalmente a me piace lasciare queste righe vuote a demarcare frasi che ritengo importanti. Inoltre mi sembra che spezzino meglio il discorso e diano un certo ritmo alla lettura. Resta comunque un'opinione molto personale, quindi sono molto flessibile in merito. Per il momento terrei questo suggerimento in sospeso. Se qualcun altro preferisse quanto suggerito però sarei ben felice di applicare le modifiche suggerite.

> Una valanga travolgente di innovazione che non da segni di rallentamento.

Reviewer 2: Credo che qui si entri nel reame delle preferenze personali.. Essendo un capitolo steso in prima battuta da @lmammino , non essendo un errore ma, appunto, una preferenza, reputo possa rimanere così.. Dal mio punto di vista queste differenze danno carattere al libro.

In questo contesto cosí dinamico, cosa significa essre uno sviluppatore "senior"?

Se dovessimo dare una definizione di un profilo "senior" in un'altra industria, ad esempio, l'indutria aeronautica, potremmo pensare al ruolo del capitano.

Il capitano di un aereo é un profilo che é responsabile di tutti gli aspetti del volo, del velivolo, dei passeggeri e del personale di bordo. É un profilo che puo' sicuramente contare su diversi anni di studio, una grande esperienza (esplicitamente espressa in "ore o miglia di volo"). É sicuramente un percorso lungo che richiede tanta dedizione, ma anche un percorso piuttosto chiaro e standardizzato e che probabilmente non ha avuto significative variazioni negli ultimi 40 anni...

Possiamo trovare una definizione simile anche in ambito software?

_Dipende!_

In questo capitolo cercheremo di esplorare cosa significa essere senior in ambito software. Discuteremo cosa puó e deve offrire un senior al suo team e alla sua azienda. Discuteremo sia aspetti tecnici (hard skills) che aspetti meno tecnici (soft skills).

> Ma prima di imbarcarci in questo viaggio, é importante puntializzare che non esiste una definizione canonica di "senior". Ogni azienda ha la sua definizione, ogni team ha la sua definizione, ogni individuo ha la sua definizione.

Reviewer 2: ```suggestion
Ma prima di imbarcarci in questo viaggio, é importante puntializzare che non esiste una definizione canonica di "senior". Ogni azienda, ogni team e ogni individuo ha la propria definizione.
```

Per evitare ripetizioni

La nostra esplorazione sará contaminata dalle opinioni e le esperienze personali degli autori e lo spirito di questo capitolo non é quello di offrire una guida assolutamente corretta su come diventare senior in 10 mosse, ma piuttosto essere un piccolo almanacco di buone pratiche per ispirare coloro che ambiscono a diventare senior e per coloro che sono giá riconosciuti come tali e sono alla ricerca di spunti per migliorarsi.

## Senior? Si, ma quanto?

> Immaginiamo per un attimo di essere il CTO di un'azienda in forte crescita. Abbiamo costruito un prodotto grazie al supporto di un'azienda di consulenza esterna, ma adesso é fondamentale portare lo sviluppo in-house e costruire un team interno per poter rendere l'azienda stabile e raggiungere i prossimi obiettivi di crescita.

Reviewer 2: ```suggestion
Immaginiamo per un attimo di essere il CTO di un'azienda in forte crescita. Abbiamo costruito un prodotto grazie al supporto di un'azienda di consulenza esterna, ma adesso é fondamentale portare lo sviluppo _in-house_ e costruire un team interno per poter rendere l'azienda stabile e raggiungere i prossimi obiettivi di crescita.
```

Ho messo in corsivo il termine essendo un inglesismo.

Quello che ci serve in questa fase é un team di sviluppatori senior che possa essere la base attorno alla quale costruire e far crescere un team piú ampio e variegato di professionisti del software.

> Iniziamo a pensare a come scrivere una job spec per la ricerca di questi sviluppatori senior. Cosa dovremmo scrivere? Cosa dovremmo cercare? Come definiamo il profilo che stiamo cercando?

Reviewer 2: ```suggestion
Iniziamo a pensare a come scrivere una _job spec_ - ovvero una descrizione dell'offerta e del ruolo - per la ricerca di questi sviluppatori senior. Cosa dovremmo scrivere? Cosa dovremmo cercare? Come definiamo il profilo che stiamo cercando?
```

Essendo un inglesismo specifico credo abbia senso spiegarlo.

> Non abbiamo ancora scritto neanche il titolo del profilo ricercato e siamo giá in preda ad un blocco creativo. Forse é il caso di andare a sbirciare cosa fanno le aziende concorrenti o altre aziende leader nel mercato del softare...

Reviewer 2: ```suggestion
Non abbiamo ancora scritto neanche il titolo del profilo ricercato e siamo giá in preda ad un blocco creativo. Forse é il caso di andare a sbirciare cosa fanno le aziende concorrenti o altre aziende leader nel mercato del software...
```

> Dopo una breve ricerca online, ci troviamo ad avere compilato una lista di possibili ruoli che sembrano essere sinonimi di "senior": Sviluppatore Senior, Sviluppatore _Staff_, Sviluppatore _Principal_, Architetto del software, Sviluppatore Mago, Sviluppatore Barbabianca, Sviluppatore Unicorno, Rockstar, Ninja, Guru...

Reviewer 2: ```suggestion
Dopo una breve ricerca online, ci troviamo ad avere compilata una lista di possibili ruoli che sembrano essere sinonimi di _senior_: Sviluppatore Senior, Sviluppatore _Staff_, Sviluppatore _Principal_, Architetto del software, Sviluppatore Mago, Sviluppatore Barbabianca, Sviluppatore Unicorno, Rockstar, Ninja, Guru...
```

Non vogliamo un eroe solitario in grado di scrivere migliaia di righe di codice al giorno e di produrre tanto quanto 10 svilupattori! Piuttosto cerchiamo un individuo che sappia giocare di squadra e che sappia valorizzare e massimizzare le competenze e le esperienze di tutti i membri del team.

Ok, facile a dirsi... ma, nella pratica, di quali competenze, compiti e responsabilitá stiamo parlando? Quale dovrebbe essere l'attitudine di questo profilo senior?

## Gli anni di esperienza

Una delle caratteristiche chiave che é possibile riscontrare nella maggior parte delle offerte di lavoro nel mercato del software é il numero di anni di esperienza.

"Ricerchiamo professionista volenteroso con 5 anni di esperienza con React.js".

Oppure piú genericamente: "Ricerchiamo professionista con almeno 10 anni di esperienza nel settore dello sviluppo software".

Questo tipo di frasi sono cosí comuni da sembrare quasi banali e ogni tanto riescono anche a rubare qualche sorriso.

> Il mio esempio preferito é quello del creatore del famoso web framework per Python FastAPI. Sebastián Ramirez (@tiangolo) postó [un tweet](https://twitter.com/tiangolo/status/1281946592459853830) menzionando ad annuncio di lavoro che cercava un professionista con 5 anni di esperienza specificamente con FastAPI. Peccato che lui aveva creato FastAPI meno di 2 anni prima!

Reviewer 2: ```suggestion
Il mio esempio preferito é quello del creatore del famoso web framework per Python FastAPI. Sebastián Ramirez (@tiangolo) postó [un tweet](https://twitter.com/tiangolo/status/1281946592459853830) menzionando annuncio di lavoro nel quale si cercava un professionista con 5 anni di esperienza specificamente con FastAPI. Peccato che lui avesse creato FastAPI meno di 2 anni prima!
```

Ah, se solo potessimo tornare indietro nel tempo!

Ok, tornando al punto, quanti anni servono per essere un senior? O ancora, qual é l'etá minima richiesta per poter essere considerati degli sviluppatori senior?

> Queste sono due delle domande piú inutili del mondo software, per quanto incredibilmente comuni!

Author: Concordo. Fuorviante mi piace di più. Modifico appena posso 🙏

> Queste sono due delle domande piú inutili del mondo software, per quanto incredibilmente comuni!

Author: ```suggestion
Queste sono due delle domande piú fuorvianti del mondo software, per quanto incredibilmente comuni!
```

Perché?

Beh, ritornando a indossare gli abiti del nostro fantomatico CTO, assumeremmo qualcuno solo perché é stato 10 anni seduto ad una scrivania a guardare lo schermo di un computer o lo assumeremmo perché é in grado di contribuire alla risoluzione di specifici problemi di business?

É vero che aver investito piú tempo in certe attivitá potrebbe significare aver avuto piú occasioni per imparare ed eccellere in una data disciplina. Ma il mondo software é una disciplina cosí complessa ed in continuo cambiamento in cui l'equazione "tanto tempo = grandi competenze" diventa una semplificazione troppo banale e rischiosa.

> Nella mia carriera ho avuto la fortuna di lavorare con dei colleghi giovani ma brillanti, in grado di comprendere a fondo problemi tecnici e di business e in grado di dare un grandissimo contributo al team in cui lavoravono.

Reviewer 2: ```suggestion
Nella mia carriera ho avuto la fortuna di lavorare con dei colleghi giovani ma brillanti, in grado di comprendere a fondo problemi tecnici e di business e in grado di dare un grandissimo contributo al team in cui lavoravano.
```

Mi é anche capitato di lavorare con colleghi con tanti anni di esperienza e anche grosse aziende sul CV che invece si sono limitati a fare il minimo indispensabile e, quando coinvolti nella risoluzione di problemi complessi, mostravano serie lacune sia tecniche che di comprensione del business.

> Certo, questi sono esempi estremi e non sono sicuramente la norma. Il punto é che non bisogna assolutamente puntare ne agli anni di esperienza ne all'etá, ma bisogna piuttosto mirare alle competenze acquisite e alla capacitá (e volontá) di continuare ad acquisirne di nuove.

Reviewer 2: ```suggestion
Certo, questi sono esempi estremi e non sono sicuramente la norma. Il punto é che non bisogna assolutamente puntare né agli anni di esperienza né all'età, ma piuttosto mirare alle competenze acquisite e alla capacità (e volontà) di continuare ad acquisirne di nuove.
```

## Le competenze

Se non é una questione di anni di esperienza sul campo su cosa dobbiamo concentrarci?

Se definiamo un senior come qualcuno in grado far progredire progetti e persone, ci saranno sicuramente delle competenze necessarie al fine di poter raggiungere questi obiettivi.

> Queste competenze non sono esclusivamente tecniche ma anche non-tecniche o di comunicazione. Uno sviluppaore senior deve essere familiare con piú linguaggi di programmazione, framework, database, architetture software, etc. Ma deve anche avere ottime capacitá comunicative, capacitá di leadership, capacitá di business e problem-solving.

Reviewer 2: ```suggestion
Queste competenze non sono esclusivamente tecniche ma anche non-tecniche o di comunicazione. Uno sviluppatore senior deve essere familiare con piú linguaggi di programmazione, framework, database, architetture software, etc. Ma deve anche avere ottime capacitá comunicative, capacitá di leadership, capacitá di business e problem-solving.
```

> A volte le skill non tecniche possono risultare anche piú importanti di quelle prettamente tecniche. Perché non importa solamente quello che sai, importa anche come tu sia in grado di trasmetterlo permettendo agli altri di sfruttare il tuo sapere. Come condividere la tua esperienza, sbloccare situazioni difficili e rendere il gruppo di lavoro autosufficiente e produttivo.

Reviewer 2: ```suggestion
A volte le _skill_ non tecniche possono risultare anche piú importanti di quelle prettamente tecniche. Perché non importa solamente quello che sai, importa anche come tu sia in grado di trasmetterlo permettendo agli altri di sfruttare il tuo sapere. Come condividere la tua esperienza, sbloccare situazioni difficili e rendere il gruppo di lavoro autosufficiente e produttivo.
```

Un’altra capacità indispensabile per lo sviluppatore senior è la propensione alla risoluzione dei problemi. Uno sviluppatore senior deve essere in grado di pensare in modo critico, analizzare problemi complessi e trovare soluzioni creative. Dovrebbe essere a suo agio all’interno di situazioni poco chiare e non ancora definite, pronto ad adattarsi ad ogni nuovo sviluppo.

In fin dei conti, l’ingegneria del software non è altro che risoluzione di problemi!

Tuttavia i problemi non sono sempre di natura tecnica, specie all’interno delle organizzazioni più strutturate, dove sono presenti interlocutori con esperienze diverse e di differente formazione, e i problemi possono essere comunicati nel modo sbagliato e mutare di frequente. Per questo diventa fondamentale contribuire alla comprensione dei problemi che si stanno riscontrando, formulare e condividere gli obiettivi e guidare la squadra di lavoro nella giusta direzione; competenze cruciali, queste, che possono avere un impatto enorme.

La migliore soluzione da un punto di vista strettamente tecnico non è necessariamente la migliore soluzione per una data azienda.

> Un senior dovrebbe essere in grado di riconoscere queste situazioni e di abbandonare soluzioni tecnicamente ottimnali in favore di soluzioni piú adatte alla squadra e alle circostanze di business.

Reviewer 2: ```suggestion
Un senior dovrebbe essere in grado di riconoscere queste situazioni e di abbandonare soluzioni tecnicamente ottimali in favore di soluzioni piú adatte alla squadra e alle circostanze di business.
```

Un esempio estremo potrebbe essere quello di uno sviluppatore capace di scrivere codice di basso livello e iper-ottimizzato per un dato obiettivo. Questo approccio potrebbe limitare l'accesso al progretto a pochissimi individui con competenze simili. Quante aziende hanno veramente l'esigenza di produrre codice cosí ottimizzato? Spesso é piú conveniente puntare all'agilitá e alla possibilitá di iterare e sperimentare velocemente.

> Le competenze tecniche da sole non bastano a comprendere tutte le sfaccettature di un progetto e a trovare soluzioni che soddisfano tutte le necessità di business e spesso i migliori compromessi possono essere trovati con il giusto equilibrio di competenze sia tecniche che di business.

Reviewer 2: ```suggestion
Le competenze tecniche da sole non bastano a comprendere tutte le sfaccettature di un progetto e a trovare soluzioni che soddisfino tutte le necessità di business e spesso i migliori compromessi possono essere trovati con il giusto equilibrio di competenze sia tecniche che di business.
```

Questo tipo di profilo permette al singolo individuo di poter contribuire in più modi ad un progetto pur essendo esperto nella propria area di specializzazione primaria. I profili a T consentono una collaborazione efficace con membri di lavoro aventi competenze diverse, aprendo la strada ad un processo di sviluppo con maggiori probabilità di successo.

Un buon modo per costruire il profilo a T è quello di iniziare a sviluppare il tuo tratto verticale, ovvero quello che alla fine diventerà il tuo punto di forza! Concentrati per un po’ su una singola area. Ovviamente é meglio preferire delle aree tecniche per le quale ci si senta particolarmente propoensi e per le quali si abbia un grosso interesse. Piú qualcosa ci appassiona, piú é facile stare al passo con l'innovazione e trovare il tempo per sperimentare ed imparare.

> Quando pensi di avere raggiunto la padronanza di una data competenza tecnica, allora potrai espandere la tua conoscenza partendo da lì, sviluppando cosí il tratto orizzontale. Esci fuori dalla tua comfort zone ed esplora argomenti correlati. Questi argomenti dovrebbero auspicabilmente arricchire le tue conoscenze pregresse e permetterti di contribuire in modo piú esteso ai vari aspetti dello sviluppo di un prodotto software.

Author: Cioé? Intendi che ci si potrebbe annoiare a fare sempre le stesse cose?

Se, ad esempio, sei diventato esperto in sviluppo di API REST, potresti voler approfondire i database o lo sviluppo front-end, in quanto l'insieme di queste competenze ti permetterebbe di avvicinarti ad un profilo da programmatore full-stack. Allo stesso modo, potresti anche approfondire l’argomento IaC (infrastructure-as-code) ed imparare ad implementarlo nel processo produttivo. Non limirarti a scrivere software ma cerca di capire anche cosa é necessario per portare quel software in produzione.

> Il segreto non sta nel diventare un esperto in ognuna di queste ulteriori aree ma nell’apprendere quanto basta per capire le basi e per poter collaborare in modo efficace con chi, invece, le padroneggia a fondo.

Author: Potresti espandere un po' su questo suggerimento?

## Comprendere i compromessi

Ad uno sviluppatore senior serve una vasta conoscenza dei principi dello sviluppo di software, che comprende non solo algoritmi e strutture dati ma anche diversi modelli architetturali, organizzazione del codice, testing, processi di sviluppo e scalabilità.

L'idea é quella di sviluppare un pensiero critico, grazie al quale si possa essere in grado di comprendere e confrontare tecnologie e paradigmi diversi e i compromessi che ne derivano.

Un esempio classico potrebbe essere quello di comprendere e discutere le differenze e i compromessi tra architetture monolitiche o a microservizi. O ancora i compromessi che derivano da scelte di ottimizzazione che favoriscono la memoria piuttosto che la CPU. E cosí via: "alta scalabilità" contro "bassa latenza", "software generico e riutilizzabile" contro "software su misura", "complesso ed efficiente" contro "semplice e flessibile".

> Non ci sono soluzioni corrette o sbagliate, ma solo compromessi differenti. Ed é fondamentale comprendere questi compromessi e prevederne l’effetto, sia nel breve che nel lungo termine. La domanda da farsi é "quale compromesso é quello migliore per gli obiettivi di business presenti e futuri"?

Reviewer 4: ```suggestion
Non ci sono soluzioni corrette o sbagliate, ma solo scelte differenti. Ed é fondamentale comprendere questi compromessi e prevederne l’effetto, sia nel breve che nel lungo termine. La domanda da farsi é "quale compromesso é quello migliore per gli obiettivi di business presenti e futuri"?
```

Author: Qui mi piace di piú "compromessi" (che ripetiamo pure nella frase successiva)...

> Non ci sono soluzioni corrette e soluzioni sbagliate, ci sono solo diversi compromessi. Ed é fondamentale comprendere questi compromessi e prevedere l’effetto, sia nel breve che nel lungo termine, dei compromessi fatti. La domanda da farsi é "quale compromesso é quello migliore per gli obiettivi di business presenti e futuri"?

Reviewer 2: ```suggestion
Non ci sono soluzioni corrette o sbagliate, ma solo compromessi differenti. Ed é fondamentale comprendere questi compromessi e prevederne l’effetto, sia nel breve che nel lungo termine. La domanda da farsi é "quale compromesso é quello migliore per gli obiettivi di business presenti e futuri"?
```

Purtroppo non esiste una formula magica per la tecnologia!

Non esiste un singolo linguaggio o framework che possa essere utilizzato per costruire qualsiasi tipo di applicazione in modo ottimale! Se hai una visione sufficientemente completa del panorama tecnologico, dovresti essere in grado di capire quale insieme di tecnologie è il piú adatto per affrontare un problema specifico. Talvolta questo significa essere in grado di mettere da parte le nostre tecnologie preferite o quelle che fanno parte delle nostre competenze primarie e che ci fanno sentire piú produttivi a livello individuale.

Ovviamente, non sempre possiamo avere una visione cristallina di ogni problema. Spesso bisogna confermare le proprie assunzioni con degli esperimenti ad hoc ed essere pronti ad essere smentiti.

Bisogna trovare il modo di lasciare margine per commettere errori, imparare dai fallimenti e provare approcci alternativi. La strada per il successo raramente è una linea retta…

## Andare a caccia di bug

È risaputo che il software privo di bug non esiste, quindi non proviamo neanche a pensare che uno sviluppatore senior possa scrivere codice senza bug!

> Benché le bug sono sempre dietro l'angolo, uno sviluppatore senior deve sapere cosa fare per ridurre al minimo il rischio di creare dei bug. Analogamente lo sviluppatore senior deve sapere come individuare i bug il più presto possibile, possibilmente prima che lo facciano gli utenti.

Reviewer 2: ```suggestion
Benché i bug sono sempre dietro l'angolo, uno sviluppatore senior deve sapere cosa fare per ridurre al minimo il rischio di creare dei bug. Analogamente lo sviluppatore senior deve sapere come individuare i bug il più presto possibile, possibilmente prima che lo facciano gli utenti.
```

> Questo comporta sviluppare tutta una serie di competenze quali: comprendere ed approfondire i resoconti degli utenti, scrivere diversi tipi di test (unitari, di integrazione, end-to-end), individuare ed analizzare i casi limite, leggere e ricercare informazioni all'interno dei log applicativi, riuscire a leggere ed interpretare varie metriche e dashboard.

Reviewer 4: ```suggestion
Questo comporta sviluppare tutta una serie di competenze quali: comprendere ed approfondire i resoconti degli utenti, scrivere diversi tipi di test (unitari, di integrazione, end-to-end), individuare ed analizzare i casi limite, strutturare i logs in mod che siano più facili da leggere e ricercare in caso di errori, riuscire a leggere ed interpretare varie metriche e dashboard.
```

## Voglia di crescere

La mentalità di uno sviluppatore senior è cruciale quanto le sue competenze tecniche. Un ingegnere senior deve avere una mentalità incline alla crescita, la volontà di imparare e adattarsi e la propensione al continuo miglioramento.

Siamo tutti consapevoli di quanto velocemente si evolva il mondo del software, lo abbiamo giá detto ed una situazione di disagio che viviamo tutti i giorni. L'unica arma in nostro possesso per poter sopravvivere in questo mondo é quella di essere in grado di imparare velocemente e di adattarci ai cambiamenti.

Imparare peró significa anche essere a proprio agio con i fallimenti e vederli come un’opportunità per imparare. Questo significa talvolta essere disposto a correre dei rischi, provare soluzioni o tecnologie con le quali non abbiamo molta confidenza ed essere pronti a tornare indietro sui propri passi qualora quell'approccio non porti i frutti sperati.

Ogni fallimento peró é un passo in avanti verso il successo, una lezione appresa che ci permetterá di non ripetere gli stessi errori in futuro e di essere piú consapevoli delle scelte che facciamo.

## Essere una leva attiva, saper dire "NO" e debito tecnico

> Uno sviluppatore senior deve avere un ruolo molto proattivo in un’organizzazione. Non dovrebbe isolarsi in una stanza e continuare a sbattere le dita contro la tastiera. Un senior deve sapere quando è il momento di fare domande difficili e prendere in mano la situazione.

Reviewer 4: "difficili" nel senso di "scomode"? Anche se "Le domande non sono mai indiscrete. Le risposte lo sono, a volte" (cit non mia)

Essere in grado di fornire un punto di vista fortemente tecnico in una conversazione aziendale può avere grande impatto e informare l’azienda su quale sia la strategia migliore per andare avanti. A tutti gli effetti, un senior dovrebbe diventare un ponte tra il prodotto e la tecnologia.

Questo a volte significa che un senior dovrebbe anche sapere quando è il momento di dire NO. No a soluzioni rapide e sporche (che alla fine si ritorceranno contro) solo per rispettare una scadenza. No a piani che prevedono solo la costruzione di funzionalità su funzionalità senza considerare l’esperienza dell’utente e la stabilità a lungo termine del prodotto. No a scelte tecniche arbitrarie solo perché “abbiamo sempre fatto così ed è sempre andata bene”.

Dire NO è facile ma al tempo stesso molto difficile. Tutti possono dire NO, questa è la parte facile! Il difficile è argomentare il perché di quel NO, proporre punti di vista alternativi, trovare compromessi e disinnescare bombe a orologeria a breve o lungo termine.

> Per riuscire a farlo serve un elevato livello di soft skills. Devi saper comunicare in modo efficace e scegliere con saggezza le tua battaglie. Non puoi dire sempre NO a tutto o corri il rischio di essere preso per un ostruzionista scorbutico.

Reviewer 4: ```suggestion
Per riuscire a farlo serve un elevato livello di soft skills: devi saper comunicare in modo efficace usando le parole giuste e scegliere con saggezza le tua battaglie. Non puoi dire sempre NO a tutto o corri il rischio di essere preso per un ostruzionista scorbutico.
```

> Si tratta piuttosto di saper dire “NO, MA” piuttosto che un secco “NO”. E dovresti anche essere in grado di farlo di fronte alle persone giuste e al momento giusto… Se fatto bene, questo può avere un forte impatto positivo sull’azienda e sul team. Può portare all’innovazione e creare nuove potenzialità di lavoro o vantaggi competitivi unici.
> 
> Ovviamente non è sempre possibile dire NO. A volte è necessario accettare compromessi e fare scelte che non sono necessariamente le migliori dal proprio punto di vista. Ma è importante essere in grado di riconoscere quando è il momento di dire NO e quando è il momento di accettare un compromesso.

Reviewer 4: ```suggestion
Si tratta piuttosto di saper dire “NO, MA” piuttosto che un secco “NO”: provvedere ad una valida alternativa, seppur non ottima tecnicamente o di requisiti è parte del ruolo. A volte è necessario accettare compromessi e fare scelte che non sono necessariamente le migliori dal proprio punto di vista. Ma è importante essere in grado di riconoscere quando è il momento di dire NO e quando è il momento di accettare un compromesso.
E dovresti anche essere in grado di farlo di fronte alle persone giuste e al momento giusto… Se fatto bene, questo può avere un forte impatto positivo sull’azienda e sul team. Può portare all’innovazione e creare nuove potenzialità di lavoro o vantaggi competitivi unici.
```

## Capire l’azienda

Per poter essere una leva attiva bisogna capire l’azienda il più profondamente possibile.

Qual è lo scopo dell’azienda? Qual è la visione nel lungo termine? Esiste una strategia chiara? Quali sono i grandi punti di forza e quali di debolezza? Chi sono i clienti? Quali sono i prodotti e i servizi offerti? Quali sono i principali concorrenti?

Un senior che sa rispondere a queste domande é in una buona posizione per utilizzare le proprie competenze tecniche per aiutare l’azienda ad avere successo.

Solo con una solida comprensione dell’azienda sarai in grado di scegliere le battaglie giuste, concentrarti su ciò che conta e di contribuire alla progettazione di sistemi che possano servire l’azienda oggi ma anche essere adattati alle esigenze future.

## La Comunicazione

> "La buona comunicazione è stimolante quanto il caffè nero e rende altrettanto difficile il prendere sonno poi" - Anne Morrow Lindbergh

> Come ingegnere del software, le capacità comunicative sono fondamentali per il tuo successo. Devi essere in grado di comunicare efficacemente con diversi interlocutori, tra cui altri ingegneri, project manager, clienti e dirigenti. Ciò richiede un’ampia gamma di abilità comunicative, dal parlare e ascoltare allo scrivere e presentare.

Reviewer 4: ```suggestion
Come ingegnere del software, le capacità comunicative sono fondamentali per il successo del riconoscimento della tua figura: devi essere in grado di comunicare efficacemente con diversi interlocutori, tra cui altri ingegneri, project manager, clienti e dirigenti. Ciò richiede un’ampia gamma di abilità comunicative, dal parlare e ascoltare allo scrivere e presentare o farle tutte e quattro assieme.
```

Poiché dovrai parlare con tutti gli interlocutori, dovresti imparare a comunicare concetti tecnici a interlocutori non tecnici in un modo che sia facile da comprendere. Ciò significa anche saper ascoltare i commenti e incorporarli nel tuo lavoro. La capacità di comunicare efficacemente con gli interlocutori è essenziale per garantire che tutti siano sulla stessa lunghezza d’onda e che i progetti siano portati a termine con successo.

Un’altra importante abilità comunicativa per un ingegnere del software è quella di spiegare le incertezze e proporre idee su come affrontarle. Lo sviluppo del software è un processo intrinsecamente incerto e ci saranno sempre incognite e sfide impreviste.

Come ingegnere del software, devi essere in grado di comunicare queste incertezze in modo chiaro e conciso. Devi anche essere in grado di proporre idee su come affrontare queste incertezze e far progredire il progetto.

> Inoltre, gli ingegneri del software devono essere in grado di parlare degli errori e delle lezioni apprese. Ne abbiamo già parlato, ma è importante sottolineare che il fallimento è una parte inevitabile dello sviluppo del software ed è importante essere trasparenti sui fallimenti e sulle lezioni apprese. Ciò richiede capacità di comunicazione efficaci, compresa quella di assumersi la responsabilità degli errori e di comunicare onestamente. Inoltre, è necessario saper comunicare le lezioni apprese dai fallimenti e incorporarle nei progetti futuri. In un certo senso, un errore puó anche essere visto come un investimento a lungo termine. Se questo viene comunicato bene puó essere percepito in modo assolutamente positivo da parte dell'azienda e puó convince a investire ancora di piú nella ricerca di nuove soluzioni e pratiche di sviluppo.

Reviewer 4: Non se ne fa cenno prima, quindi un "Ne abbiamo già parlato" mi trova spaesato, ma magari me lo sono perso io...

È bene essere chiari con la direzione e con il resto del team tecnico e definire le caratteristiche del successo individuale e di squadra. Se non sai cosa ci si aspetta da te e dal resto del team, anche se pensi di fare la cosa giusta, potresti finire per non fornire il valore che ci si aspetta da te; allo stesso modo, potresti non essere in grado di aiutare il tuo team ad avere successo.

Avere un simile atteggiamento di schiettezza e chiarezza di solito mantiene l’umore più positivo e aiuta a evitare le delusioni. Certo, non è sempre possibile evitare tutte le delusioni o i fallimenti. Perciò, quando il problema si presenta, potresti essere tentato di dare la colpa a qualcun altro o al sistema. Ma è invece più maturo chiedersi cosa si sarebbe potuto fare meglio, o ancora meglio: cosa si potrebbe fare in futuro per evitare che problemi simili si ripresentino.

## Come posso crescere?

Se sei un ingegnere del software e vuoi aumentare le tue competenze e diventare un membro senior del tuo team, potresti chiederti quali sono i passi da intraprendere per raggiungere questo obiettivo.

Beh, non temere! Ecco 4 idee (più alcune extra) che credo ti aiuteranno nel tuo percorso.

> ## Scendere giú di un livello

Reviewer 4: questi dovrebbero avere ### se non ho capito male, giusto?

Tutti questi sono argomenti che puoi esplorare per consolidare la tua conoscenza di questo particolare ambito. Ti permetteranno di esplorare concetti più generali che potresti riutilizzare in altri ambiti. Tutti costituiscono dei validi passi in avanti per ampliare le tue conoscenze e diventare un ingegnere più completo.

Ed il modo migliore per esplorare questi piani é quello di costruire qualcosa. Puoi capire veramente qualcosa se riesci a costruirne un piccolo prototipo. Ad esempio potresti provare a decodificare un token JWT senza usare una libreria. Saresti poi in grado di andare oltre e implementare anche l’algoritmo di verifica della firma? Magari poresti anche provare a implementare un server OAuth2.0 per capire meglio come funziona il protocollo.

Costruire prototipi è un ottimo modo per memorizzare alcuni concetti e mettere davvero alla prova la vostra conoscenza. Naturalmente, costruire richiede uno sforzo maggiore, quindi scegli con saggezza gli esercizi di prototipazione: non è possibile reimplementare oltre 60 anni di ingegneria del software solo per il gusto di imparare!

> Un video divertente che mi è piaciuto molto ultimamente è [The Computer Science Iceberg](https://www.youtube.com/watch?v=H565avw-ufk). Perché mi piace così tanto? Perché é diverte, ironico e soprattutto perché illustra davvero bene l’idea di livelli graduali di astrazione e di ricerca di conoscenze fondamentali e generiche.

Reviewer 2: ```suggestion
Un video divertente che mi è piaciuto molto ultimamente è [The Computer Science Iceberg](https://www.youtube.com/watch?v=H565avw-ufk). Perché mi piace così tanto? Perché é divertente, ironico e soprattutto perché illustra davvero bene l’idea di livelli graduali di astrazione e di ricerca di conoscenze fondamentali e generiche.
```
